Added recipe for scotch

basnijholt opened the pull request.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/scotch```) and found some lint.

Here's what I've got...


For **recipes/scotch**:

 * The recipe must have some tests.
 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/scotch```) and found some lint.

Here's what I've got...


For **recipes/scotch**:

 * The recipe must have some tests.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/scotch```) and found it was in an excellent condition.

jakirkham commented:
Could you please restore the `LICENSE` file? It seems to be overwritten by something else.

jakirkham commented on recipes/scotch/build.sh:
```diff
@@ -0,0 +1,25 @@
+  sed -i "s@CLIBFLAGS\t=@CLIBFLAGS\t= -fPIC@g" Makefile.inc
+  sed -i 's#-l$(SCOTCHLIB)errexit#-l$(SCOTCHLIB)errexit -lm#g' esmumps/Makefile
+fi
+make esmumps | tee make.log 2>&1
```
`make check` or similar?

jakirkham commented on recipes/scotch/build.sh:
```diff
@@ -0,0 +1,25 @@
+mkdir -p $PREFIX/bin/
+cp bin/* $PREFIX/bin/
+mkdir -p $PREFIX/include/
+cp include/* $PREFIX/include/
```
No `make install`?

basnijholt replied:
`make esmumps` also does `make install`

basnijholt replied:
> No make install?

`make esmumps` also does `make install` too

jakirkham replied:
So that's confusing. Why do we need to copy stuff over then?

jakirkham commented on recipes/scotch/build.sh:
```diff
@@ -0,0 +1,25 @@
+#!/bin/sh
+
+cd src/
```
`./configure`?

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+source:
+  fn: scotch_{{ version }}.tar.gz
+  md5: {{ md5 }}
+  url: {{ file_url }}
```
This shouldn't need to be templated. Please keep the URL here.

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+
+requirements:
+ build:
+  - gcc
```
Please drop and replace with `toolchain`.

jakirkham replied:
This configures compilation on the OSes to use the system toolchains. On OS X, we use `clang`. On Linux, we use gcc from devtoolset-2.

basnijholt replied:
I reversed the change again, because it didn't compile:

```
/opt/rh/devtoolset-2/root/usr/libexec/gcc/x86_64-redhat-linux/4.8.2/ld: cannot find -lz
```

jakirkham replied:
I hear that this is a problem, but this isn't the right fix. Please re-reverse it and look at [comment]( https://github.com/conda-forge/staged-recipes/pull/1427/files#r76904134 ).

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+  - zlib  # [linux]
+ run:
+  - zlib  # [linux]
+  - libgcc
```
Please drop.

jakirkham replied:
There is no need to ship `libgcc` as we are dependent only on system libraries that the system toolchains use.

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+requirements:
+ build:
+  - gcc
+  - zlib  # [linux]
```
Why only on Linux?

jakirkham replied:
Please change to `zlib 1.2.*` also.

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+  - gcc
+  - zlib  # [linux]
+ run:
+  - zlib  # [linux]
```
Please change to `zlib 1.2.*`.

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+  recipe-maintainers:
+    - basnijholt
+
+
```
nit: Please strip extra newlines. One terminal newline should be sufficient.

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+
+test:
+  commands:
+    - test  # [osx]
```
Can we check to make sure all files are installed correctly? Is there a CLI that we can use to run `--help`, `--version`, or similar?

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+  md5: {{ md5 }}
+  url: {{ file_url }}
+  patches:
+    - clock_gettime.patch  # [osx]
```
What's the purpose of this patch? Has it been submitted upstream?

basnijholt replied:
To be honest I don't know. It was done in another recipe I found and on the homebrew version.

jakirkham replied:
This [one]( https://github.com/Homebrew/homebrew-science/blob/ce9c31eea5f42c0b894a8eb2a24bd17bc7880c7b/scotch.rb ). If so, I'm not seeing the patch. I did see a similarly named patch for Perl. In any event, I think we can drop this.

jakirkham commented on recipes/scotch/build.sh:
```diff
@@ -0,0 +1,26 @@
+    sed -i '' 's/-O3/-O3 -fPIC/g' Makefile.inc
+else
+  cp Make.inc/Makefile.inc.x86-64_pc_linux2 Makefile.inc
+  sed -i "s@CFLAGS\t\t=@CFLAGS\t= -I${PREFIX}/include@" Makefile.inc
```
Might need to do this for the library search path too.

basnijholt replied:
I don't know why or how?

jakirkham replied:
Why would be to pick up `zlib`.

How would be to add `-L${PREFIX}/lib`. Though I really am hoping they have a configure file we overlooked.

jakirkham replied:
So I looked at the source briefly. They don't have a configure. 😖

My recommendation would be to copy these files into the recipe and modify them as you see fit. We did this with [SuiteSparse]( https://github.com/conda-forge/suitesparse-feedstock/tree/41f46eef1e79839154a7211460a883b3478228fa/recipe ) and had much success. Keeping this `sed` stuff is more likely to cause you hurt later on.

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+source:
+  fn: scotch_{{ version }}.tar.gz
+  md5: {{ md5 }}
+  url: http://gforge.inria.fr/frs/download.php/file/34618/scotch_6.0.4.tar.gz
```
`scotch_6.0.4.tar.gz` -> `scotch_{{ version }}.tar.gz`. That should make it easier to update. 😉

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+requirements:
+ build:
+  - gcc
+  - zlib 1.2.*  # [linux]
```
Please change to `unix`.

jakirkham replied:
We also have `bison` and `flex`. It looks like they would come in handy here.

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,46 @@
+  - gcc
+  - zlib 1.2.*  # [linux]
+ run:
+  - zlib 1.2.*  # [linux]
```
Please change to `unix`.

basnijholt commented:
Thanks a lot @jakirkham!

Why no `zlib` for OSX, I don't remember, I (and others in my group) have worked with the compiled version of scotch for a while now and it has always worked fine.

The only remaining thing is a test and in the documentation I can't find a way to test it in command line interface. However, during the compilation we do do `make check`, which succeeds.

basnijholt commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+    - test -f "${PREFIX}/lib/libscotcherr.a"
+    - test -f "${PREFIX}/lib/libscotcherrexit.a"
+    - test -f "${PREFIX}/lib/libscotchmetis.a"
+    - test -f "${PREFIX}/lib/libesmumps.a"
```
@jakirkham I can't find a better way to test these files? What do you think?

jakirkham replied:
Can we check for some dylibs/sos? Also some tests for installed headers would be helpful.

basnijholt replied:
There are no dylibs/sos as far as I can see. I did include the header file tests now.

jakirkham replied:
Yeah, I'm not seeing them either now that I'm looking. Is there an option to build dynamic libraries?

shadowwalkersb replied:
@jakirkham , it looks like the Makefiles are written by hand. Thus, no `./configure` (regarding your question above). Might have to add the flags manually. `-shared` may do the trick. Sorry, if this is already addressed somewhere in the conversations.

jakirkham commented on recipes/scotch/Makefile.inc.i686_mac_darwin10:
```diff
@@ -0,0 +1,21 @@
+CAT		= cat
+CCS		= gcc
+CCP		= mpicc
+CCD		= gcc
```
Why not use `clang`?

basnijholt replied:
Done :)

basnijholt replied:
Should I do it in the Linux `Makefile.inc` too?

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+  version: {{ version }}
+
+source:
+  fn: scotch_6.0.4.tar.gz
```
Please change to `{{ name }}-{{ version }}.tar.gz`.

basnijholt replied:
Done

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,50 @@
+source:
+  fn: scotch_6.0.4.tar.gz
+  md5: {{ md5 }}
+  url: http://gforge.inria.fr/frs/download.php/file/34618/scotch_6.0.4.tar.gz
```
Please template the version with jinja here.

basnijholt replied:
I changed it, but because the number in the url changes for every file I initially didn't.

basnijholt commented:
@jakirkham I've changed what you suggested, two things left:
> Can we check for some dylibs/sos?

I don't see that there are any dylibs or sos files generated

>Why not use clang?

 I have changed it for the MacOS makefile, but should I do it for Linux too?

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,54 @@
+source:
+  fn: {{ name }}-{{ version }}.tar.gz
+  md5: {{ md5 }}
+  url: http://gforge.inria.fr/frs/download.php/file/34618/{{ name }}-{{ version }}.tar.gz
```
Could try the link below. The number on it doesn't seem to change with each release. So it works better with templating.

```
https://gforge.inria.fr/frs/download.php/latestfile/298/{{ name }}_{{ version }}.tar.gz
```

jakirkham commented:
Seems to be picking up the [system `zlib`]( https://travis-ci.org/conda-forge/staged-recipes/builds/162395113#L1514 ). Could we fix it to pick up our `zlib`?

jakirkham commented on recipes/scotch/meta.yaml:
```diff
@@ -0,0 +1,54 @@
+  - zlib 1.2.*  # [unix]
+
+test:
+  commands:
```
There seems to be a bunch of executables that are installed. Can we please run `--help` or `--version` or similar to check that they do run without crashing?

basnijholt replied:
I have added a few.

basnijholt commented:
I have changed everything @jakirkham suggested, except that I don't know how to generate dynamic libraries, the `-fPIC` flag is already there at least.

basnijholt commented:
@jakirkham I've checked there is a Makefile.inc for Linux, not for OSX. However, it's not compiling correctly (d87384b), so I reverted the change.

basnijholt commented:
I have spend a lot of time to get the dynamic lib compilation to work, but it's not working. 

Also by looking for info about the `Makefile.inc.x86-64_pc_linux2.shlib` online, doesn't give me the confidence of being sure that it works at all.

shadowwalkersb commented:
Try `-dynamiclib` for dynamic libraries.

In my cmake generated installation command I also see the the flags below. Maybe not needed, but I am listing them just in case.

` -Wl,-headerpad_max_install_names`
`-install_name`
`-framework`

jakirkham commented:
No worries. I think we are now at a healthy place to merge. While there is still room for growth, I'm more than happy to see that continue at the feedstock. Thanks very much for your hard work, @basnijholt .
